Add unit tests to remove instance button component

AlexSCorey opened the pull request:
##### SUMMARY
This adds UI unit tests the the new Remove Instance Button component.

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - UI

AlexSCorey committed:
Adds unit tests for RemoveInstanceButton

kialam reviewed (changes requested):
There are several things to point out:
1. Mixing Enzyme and RTL methods in `enzymeHelpers.js` seems counter-intuitive. If we need to create a context mapper/render using RTL since we are moving away from Enzyme, then that should be a separate method in a separate file. Maybe called `rtlHelpers.js` or something to that effect.
2. Components being used in Screens often do not need to be mounted with any React Context Providers and can (and should) be tested in isolation.
3. To move this forward I would remove the changes to `enzymeHelpers.js` and adjust the first test case to be rendered without Contexts.

AlexSCorey commented:
> There are several things to point out:
> 
> 1. Mixing Enzyme and RTL methods in `enzymeHelpers.js` seems counter-intuitive. If we need to create a context mapper/render using RTL since we are moving away from Enzyme, then that should be a separate method in a separate file. Maybe called `rtlHelpers.js` or something to that effect.

I've removed the changes in this file.  If we decide we want a help function for contexts with rtl tests we can add it again at a later time.  I do not suspect it will have a significant number of use cases.
> 2. Components being used in Screens often do not need to be mounted with any React Context Providers and can (and should) be tested in isolation.

The first test in that file does need the i18n context so that when the modal opens, in the test, the `<Plural/>` component can handle the strings. The other tests do not need it.  


> 3. To move this forward I would remove the changes to `enzymeHelpers.js` and adjust the first test case to be rendered without Contexts.

Removed.

kialam commented on awx/ui/src/screens/Instances/Shared/RemoveInstanceButton.test.js:
```diff
@@ -0,0 +1,133 @@
+    );
+  });
+
+  test('Should handle error when fetching warning message details.', async () => {
```
If you want to test how we handle PATCH/POST errors after the user has clicked the button, then I would rewrite this test as the following:
1. Attempt to remove an instance (e.g. click the button)
2. Mock the `InstancesAPI` return value to give an Error
3. Check that the error shows up in the UI

AlexSCorey replied:
This specific test is ensuring that we handle errors when we are gathering info about instance_groups that use the instance to be deprovisioned.

kialam commented on awx/ui/src/screens/Instances/Shared/RemoveInstanceButton.test.js:
```diff
@@ -0,0 +1,133 @@
+    const user = userEvent.setup();
+    render(
+      <RemoveInstanceButton
+        isK8s={true}
```
I might be missing something here but if the remove button is disabled then shouldn't `isK8s === false`? Or is there some other condition that allows this test to pass?

AlexSCorey replied:
In this case, the button is disabled because the selected instance to be removed is `node_type === control`

kialam commented on awx/ui/src/screens/Instances/Shared/RemoveInstanceButton.test.js:
```diff
@@ -0,0 +1,133 @@
+  },
+];
+describe('<RemoveInstanceButtton />', () => {
+  test('Should open modal and deprovision node', async () => {
```
This is an excellent test case. I like how you simulated a user interaction with the component being tested and wrote your assertions to check for relevant component state updates and function callbacks.

kialam reviewed (approved):
These tests look good. Nice work.
